### ISSUE TITLE
Play Modes Question

### ISSUE THREAD

**Issue by LocutusOfPenguin** [jromang/picochess#91]
In the start pos you can change the PlayModes by wQ from A5-D5.
The documentation is out of date with that.

For my future software development, i need to know:
What does exactly each mode (should) do?

I also looked in the old V022 code, but there is only Analysis & Game Mode implemented.
Not all modes are - for me - equal important. i personally don't need something like "book mode (a5)" see: documentation.
Therefore i want to implement the "Analysis Mode" but need some feedback.

**Comment by sshivaji** [comment]
No need for book mode, you can always get “book mode” by pressing the hint button.

Analysis mode makes sense. It might be as simple as either observe mode where pressing a clock button gives you the best move and keeps updating?

**Comment by LocutusOfPenguin** [comment]
I'm still confused.
Analysis mode should be implemented as on V022. Something like depth, value, .... in rotation or so. That's more or less clear. Also for that the documentation & old code is valid.

But what is the difference to "observe" & "kibitz"? I still don't understand , sorry ;-)

For book-mode OK. That's fine.

**Comment by LocutusOfPenguin** [comment]
On my branch
https://github.com/LocutusOfPenguin/picochess/tree/PythonChessV0.8
i uploaded a working analysis version.

For the moment you can see move-hint on button 2, and value on button 4 (=no rotation!).
The old Button 4 stuff, i moved to a submenu (for what do i need that? The version is displayed on beginning).

Also there is a message if picochess is still in book and other small improvements.

My last comment is still valid => difference?

**Comment by djdekker** [comment]
Kibitz mode seems to be the same as Analysis mode. After all, a kibitzer is not playing but only commenting, just like someone analysing a game.
In Observe mode the clock currently displays the last move. This seems redundant to me. It would be better if the clock displayed the remaining time for both players (which could be set with the white queen just like in Game mode).

**Comment by ifekali** [comment]
> On 5. jun. 2015, at 08.55, Jürgen Précour <notifications@github.com> wrote:
> 
> i uploaded a working analysis version.

I'm looking at it, testing ... How do you enter Analysis mode?

> 
> For the moment you can see move-hint on button 2, and value on button 4 (=no rotation!).

Are this the third and the last button?

> The old Button 4 stuff, i moved to a submenu (for what do i need that? The version is displayed on beginning).
> 

Where? I don't see anything ...

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by ifekali** [comment]
Oh, I forgot, there's no entering via menus, only via White Queen ...

> On 5. jun. 2015, at 15.35, Jonas Žnidaršič <jonas@jonas.si> wrote:
> 
> I'm looking at it, testing ... How do you enter Analysis mode?

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by ifekali** [comment]
> On 5. jun. 2015, at 15.35, Jonas Žnidaršič <jonas@jonas.si> wrote:
> 
> I'm looking at it, testing

Ok, the second and the fourth button are working 
My notes:

Analysis mode is over when starting position is set up, requiring to put white queen on b5 again. (this is maybe a drag)
Takeback doesn't work (kinda limiting the analysis). If you take back the takeback, (going back to position before takeback) countdown starts.
What's with the "c" in eval display? :)

As it is, why is the Analysis mode needed? Wouldn't it be just fine if second and fourth button displayd hint and eval in normal Game mode?

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by ifekali** [comment]
> On 5. jun. 2015, at 15.58, Jonas Žnidaršič <jonas@jonas.si> wrote:
> 
> As it is, why is the Analysis mode needed? 

I mean, isn't this same as Observe (or Kibitz) mode? (And yes, hint and eval would be welcome in normal Game mode too, I think.)


- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by LocutusOfPenguin** [comment]
Buttons "-" & "+"

And right now, only at start pos with wQ (i gonna change that soon!). But i found an error at setup-position.

Surely, the buttons also working at game mode...But i understood, on analytics mode, there should be a rotation (which makes sense). But i don't know, how to implement a 2sec rotation in python right now, ha.

**Comment by LocutusOfPenguin** [comment]
Jonas: Well, that was the basic Question.
What this modes should do? For me, i only need something like analytics, where i see the depth, hint and value (like on other chess computers).
I would cancel wQc5, kibitz, observe, and so on...most not working, others no need (like Qc5 = white)

**Comment by sshivaji** [comment]
I thin Jurgen's aim is to simulate infinite analysis mode, instead of just getting a hint. Jurgen, go ahead and submit a pull request. I will merge in and we can tweak it.

What do you mean by 2 second rotation, just refresh every 2 seconds?

I apologize that I cannot easily put an umlaut on your name when using the ipad.

**Comment by ifekali** [comment]
> On 5. jun. 2015, at 16.27, Shivkumar Shivaji <notifications@github.com> wrote:
> 
> I thin Jurgen's aim is to simulate infinite analysis mode, instead of just getting a hint.

Ok, got it.

I think best would be to remove all non-working modes and then add them again if and when needed to avoid confusion with users. For start I propose three:

1. Game
2. Observe or Analysis (these can be basically the same, no?)
3. Position Setup (this one not set with white queen, menu only, as it is usuallly done in a non-startup position)

The hint & eval buttons should probably work in all modes. 

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by LocutusOfPenguin** [comment]
There is an issue with clock...The clock runs (at own will hmmm), going back to an old value (arggh!), and such strange things. Therefore, its on a BRANCH (right now). I'll hope to fix alot until monday.

My first name is german and my last name french..I have 1 special language sign to stress out keyboards, haha =>No problem!

On old version, you had a 2 secs rotation (see documentation).
This - if you ask me(!!) - should be the difference to play mode and analytics mode.
I propose, on play mode, press some button to get the info you always see in rotation on analytics.
Game => clocks, and Analytics => the results of Stockfish.

JÜrgen ;-)

**Comment by ifekali** [comment]
In play mode, the engine is your oponent (makes moves), while in analytics it only observes & evaluates the board (or a game between two humans).


> On 5. jun. 2015, at 16.39, Jürgen Précour <notifications@github.com> wrote:
> 
> This - if you ask me(!!) - should be the difference to play mode and analytics mode.

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by ifekali** [comment]
Rotation seems distractin and I imagine in analysis I would usually want to see evaluation most of the time. What if in analysis eval would be always displayed  but for a hint you would need to press a button? (if hint is always visible, it is kind of a spoiler for a studying player, trying to find best move, no?)

> On 5. jun. 2015, at 16.39, Jürgen Précour <notifications@github.com> wrote:
> 
> I propose, on play mode, press some button to get the info you always see in rotation on analytics.

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by sshivaji** [comment]
I will help debug over the weekend. Thanks for the progress!

**Comment by LocutusOfPenguin** [comment]
ifekali:
also a good suggest! Others would then miss the clock time ;-)

Maybe what is displayed (without a button click) can be the difference between the modes (analy, kibitz, observe)? to come BACK to the basis question from me.

**Comment by ifekali** [comment]
> On 5. jun. 2015, at 16.57, Jürgen Précour <notifications@github.com> wrote:
> 
> Maybe what is displayed (without a button click) can be the difference between the modes (analy, kibitz, observe)? to come BACK to the basis question from me.
> 
> 

hmmm, interesting ... Something like

Analysis - displays Eval by default, hint on button press
Observe - displays clock time by default, eval AND hint on button press (two buttons)
Kibitz - rotates the displays?

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by ifekali** [comment]
> On 5. jun. 2015, at 17.03, Jonas Žnidaršič <jonas@jonas.si> wrote:
> 
> Analysis - displays Eval by default, hint on button press
> Observe - displays clock time by default, eval AND hint on button press (two buttons)
> Kibitz - rotates the displays?

A better suggestion:

Analysis - displays Eval by default, hint on button press
Observe - displays clock time by default, eval AND hint on button press (two buttons)
Kibitz - displays moves as they are made and beeps, eval AND hint on button press (two buttons)

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by LocutusOfPenguin** [comment]
Alot of nice features, like "switch mode", can only be used in starting position (but i use alot function "setup"  for example for endgame training). Therefore - in my plan - i wanted to only use 1 Button for hint and the Button 4 for changing modes. Giving the info depth, move score, ... on multi clicks in rotation on button 2 (idea).

Then i have a button free (no 4) for the important function "switch modes". For sure, i can put this on a submenu somewhere, but i think, changing the modes is often used, no?

=> I like your idea, but would like to not use too much buttons (like right now on my own branch - on active development  - I did what i did cause of easy! development).

**Comment by ifekali** [comment]
I like the idea of ALL buttons being used, avoiding user's questions "what does the third button do, nothing?"

If so, a button should keep it's function in all modes for consistency, only changing it's function if absolutely needed (in a different game mode context maybe).


> On 5. jun. 2015, at 17.23, Jürgen Précour <notifications@github.com> wrote:
> 
> => I like your idea, but would like to not use too much buttons

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by LocutusOfPenguin** [comment]
OK,
many many thanks for the active discussion so far!

Cause i only have this weekend I decide the following roadmap.
If someone unhappy => shout // for sure hints/improvements very welcomed.
My focus is on having a working analyse mode.


Buttons
======
Button 2 is the hint button. It shows the best/ponder move, value and depth(?) in rotation on clicks.
This will work on all game modes.
Button 4 is right now free, and i will use it for switching modes. Lateron it can be moved to a submenu if we have a really more useful function for the standard user (not moving into deep of the upcoming menu). Implemented after function Button 2.
The rest of buttons stay the same. Btw. after this work, i will do something with the middle button, like stop search, alternative move but that's another story.


Game modes
===========
- Game (wQa5)
Usage: Play against Picochess
no change (despite Button2 working).

- Analyse (wQb5)
Usage: Go over a played game (single user)
Computer thinking while you play. The clock is NOT running. It always shows the best move.
Rest of Info can be seen with button2.
Reason: I think, the best move is what you get from a teacher too. He wont tell you ahead by 2,56324... pawns. Also this is the info, i want to see [Point] ;-) ... Surely for example a "kibitz" mode can be implemented where you constantly see something else, like a value.

- Observe (wQc5 or d5 with kibitz)
Usage: Picochess watches the game (and comment on Button2).  One human player on each side.
In contrast to Analyse here the clock IS running and always showing the time.
Reason: check the game, do pgn, and lateron can be switched to lever support for a (done) move.

- Kibitz
delete (but see above)


Hopefully you can follow my reasoning and roadmap.

**Comment by djdekker** [comment]
This is absolutely a brilliant development of Picochess! Thank you very much!!

**Comment by ifekali** [comment]
I concur.

> On 6. jun. 2015, at 10.49, DJ Dekker <notifications@github.com> wrote:
> 
> This is absolutely a brilliant development of Picochess! Thank you very much!!
> 
> 

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by sshivaji** [comment]
I agree as well. Great plan!

I am happy to test and to help with coding and debugging. If I notice improvements on your branch I can comment as well.

**Comment by ifekali** [comment]
Is there anyone willing to tackle Remote Play? (Connecting two humans together, one playing on DGT board, the other via web interface?)

I'm lobbying for this for a year now and believe this would be a killer feature, as an average USB board user bought the board for it, but soon found out it don't work with sexisting DGT software ....

With picochess enabling this, there would be no need for usernames or ICC accounts, etc., one would simply give a friend an URL of the Pico web interface and voila, the game is on!

> On 6. jun. 2015, at 15.22, Shivkumar Shivaji <notifications@github.com> wrote:
> 
> I agree as well. Great plan!

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by sshivaji** [comment]
If Jurgen is doing the button stuff, I can tackle this one. I have some improvements to the UI to commit as well.

**Comment by LocutusOfPenguin** [comment]
the observe mode working already. But when press the button, the time display stops. I need to find out, how to come back (after say 0,5sec display it)...Stay tuned... Soon update.

**Comment by ifekali** [comment]
What if a button would simply rotate the default display? If a move is displayed at all times by default (I like your reasoning), I could press a button to change it to evaluation being displayed from then on, etc. ...

> On 6. jun. 2015, at 09.57, Jürgen Précour <notifications@github.com> wrote:
> 
> 	• Rest of Info can be seen with button2.
> Reason: I think, the best move is what you get from a teacher too. He wont tell you ahead by 2,56324... pawns. Also this is the info, i want to see [Point] ;-) ... Surely for example a "kibitz" mode can be implemented where you constantly see something else, like a value.
> 
> 

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by LocutusOfPenguin** [comment]
@ifekali 
For this i wanted to use the "+" Button(4).
I implemented the rotation of Button2 now, so Button4 is now free. But i didn't had the time to do the rotation of mode changes there. Also i did a "kibitz" mode where you can see the value constantly.

There is an issue with "sliding pieces", sometimes sending a computer move, and then starting the clock :-(
Today i took a look, but cant made it fully working (in my limited time). Same for "switch sides".

Anyway, i think its stable enough and giving a huge improvement, so my branch could be merged.

**Comment by ifekali** [comment]
I am testing your latest build, very interesting. The hint moves are in long notation, you probably plan to change this to short, I assume?

I don't like rotation on the second button: I was playing a game and wanted to see the evaluation here and there and it's a drag having to push the button twice. On top of that, when I want to see the eval, I must avert my eyes on the first press, because I DON'T WANT to see the hint and cheat the machine. I only want to see how I stand.

I believe eval and hint in normal game mode should have separate buttons.

BTW, maybe we could connect via Skype and brainstorm faster sometimes? My handle is "ifekali".

1. On the dgt 3000 clock there is enough display space to  accomodate
> On 7. jun. 2015, at 23.33, Jürgen Précour <notifications@github.com> wrote:
> 
> rotation

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by sshivaji** [comment]
I merged the change. There are few minor things to fix.

It might be better to have eval first and then if you press the button the
next time, then show move?

**Comment by LocutusOfPenguin** [comment]
@sshivaji 
perfect!

The biggest issue is the "sliding move" stuff. The code there was not used in old version (abit hided) on Analyse mode, and also here it isn't (cause of the "if").
It go to the "COMPUTER_MOVE_DONE_ON_BOARD" part from time to time causing more problems.
I think, this is all ONE bug. If you can fix it, i would be very happy!!

I wasn't happy with my progress yesterday, and also i'm not with this part
of code "if case(Event.FEN)" at picochess.py
But - as said - i failed to find a better version yesterday, ha.
Now my time is very limited to post new code (for some time).

Jürgen

BTW. if you cant write "ü" , its better to use "ue" ( "ae", "oe"), cause the sound is similar, but "u" &"ü" is complete different (german rule) ;-)

**Comment by sshivaji** [comment]
I recall we can fix the sliding move issue using an approach done by Jean-Francois in the past. The idea would be to start a thread and cancel it if sliding occurred.

What don't you like about "if case (Event.FEN)"?

Thanks for the German pronunciation note,  Jürgen :)

Shiv

**Comment by LocutusOfPenguin** [comment]
@ifekali 
- Long notation: Right. Should be changed. (i often develop with XL, so i missed it).
- Skype: i send you a request. Also right, it's much easier and quicker. I like that.

**Comment by LocutusOfPenguin** [comment]
@sshivaji 
Maybe i shouldn't have said that ...Sorry.
Basically i was frustrated cause i didn't manage to work out my plan for the weekend :-) :-)

I really want this analyse, haha!


I had problems to understand what is going on.
- The line 252 wont work on analyse. I dont know how, but to find out if its a sliding move should be implemented in different way.
- there are 2 places, where a takeback it done. Is that really needed?
- compute legal fens(), i wanted to have in this function. Right now, its everywhere ... making it hard to understand (for me!)

Altogether , i have my problems understanding the various "if"'s - but that prob. my own problem! Sorry.

Jürgen

**Comment by LocutusOfPenguin** [comment]
Button stuff:
as before, we can have Button2+4 for the events "move" & "value"
I believed (in the past), that on the "position", i don't have the wQa5-d5 stuff, so i cant use the game mode change anymore.
BUT:
There the Button 4 is free, so the mode rotation can be implemented there.

Also, you can have this rotation on Button 4 for the consistency in Game mode if you happy with the change of Button 2 order propose (just stopping pressing, if you don't want to cheat).

**Comment by ifekali** [comment]
> On 8. jun. 2015, at 17.41, Jürgen Précour <notifications@github.com> wrote:
> 
> There the Button 4 is free, so the mode rotation can be implemented there.
> 
> Also, you can have this rotation on Button 4 for the consistency in Game mode if you happy with the change of Button 2 order propose (just stopping pressing, if you don't want to cheat).
> 

This could work, yes ...

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by puba** [comment]
About human remote playing on web server, I agree with Jonas that is got to be simple as pasulj, but it's got to have some kind of registration/login option, something like Sshiv's web server

**Comment by LocutusOfPenguin** [comment]
on my fork i did a small improvement with button 2 (short notation & value first).

**Comment by ifekali** [comment]
Looks good.

1. evaluation shows minus sign for the White advantage
2. if button 2 is pressed when picochess is thinking, it will show hint first, not eval (bug or a feature?)


> On 11. jun. 2015, at 12.07, Jürgen Précour <notifications@github.com> wrote:
> 
> on my fork i did a small improvement with button 2 (short notation & value first).
> 
> 

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by ifekali** [comment]
Tested some more ...

> On 11. jun. 2015, at 13.17, Jonas Žnidaršič <jonas@jonas.si> wrote:
> 
> 2. if button 2 is pressed when picochess is thinking, it will show hint first, not eval (bug or a feature?)

This happens randomly, sometimes it shows eval, sometimes (the previous) hint. But as I said in previous comment, only when machine is thinking and has not yet moved. I tested mostly with fixed time control (5 seconds to move),

3. If in the book, the evaluation will display "none". It would be probably more informative to display "book"?
4. After eval is displayed, there is stays the same even when machine has time to reevaluate. So if user hits the eval button fast, it will stay somewhat incorrect, having not much time to evaluate correctly. It should probably refresh when rotating the button, no?

Good work.

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by ifekali** [comment]
> On 11. jun. 2015, at 13.42, Jonas Žnidaršič <jonas@jonas.si> wrote:
> 
> Tested some more


5. In observe mode, the eval jumps from minus to plus depending which side is to move
6. In analysis mode, the evaluation will often not show (2nd button doesn't respond or shows hint)

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by LocutusOfPenguin** [comment]
1+5)
Its shown as the engine sends the value. So its always for the side of computer. "+" computer ahead, "-" person ahead. Even the sold like "geschnitten Brot" (hello to puba) mephistos doing so. I dont think this is a bug - but can be changed (i think the docu saying something about this). Not too important if you ask me ;-)

2) shouldnt happen. Ill take a look. Btw. Its a feature, i will update it to do something different, haha. I seldom do bugs, only features which arent working.
3) Ok
4) For what mode you talking? On Game mode, the value&hint wont change. On button click, you gonna see the ponder move, and its eval (based on the last search from computer).
The other modes, gonna change the values..but displayed only when you rotate around (for the button, display always update). Thats prob. what you mean, yes?
6) Hmmmm, ok ill see if i can do something for it

**Comment by sshivaji** [comment]
I can fix #1 (its just a one line fix) if needed - if turn to move is black, multiply eval by -1..

Shiv

**Comment by puba** [comment]
Hello Jürgen, schön dich zu lesen auch...

**Comment by LocutusOfPenguin** [comment]
I was wondering (before) what "pasulj" is. So i asked google, and therefore i used an idiom which is
"Es verkauft sich wie geschnitten Brot" [the correct english translation is "It's selling like hot cakes"]

Sadly, the german grammer is very very complex..so its often not 100% correct when translating 1:1 from english(or another language) to german, so i'm not 100% sure, what you wanted to say.

Now back to some development. Seems no 3, really working now.

**Comment by LocutusOfPenguin** [comment]
My development for this is (almost) finished.
There are some issues from ifekali left, and some are from me (like the "clock starting" problem)
but i think the issues should be on another thread after sshivaji accepted my pull-request, cause its getting hard to follow this 47 comments.

Jürgen

**Comment by ifekali** [comment]
Cool. As soon it is pulled I will test some, merge with stable and release new image over the weekend.

> On 12. jun. 2015, at 12.03, Jürgen Précour <notifications@github.com> wrote:
> 
> My development for this is (almost) finished.

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by ifekali** [comment]
When evaluating, the minus sign turning on and off (say in kibitz mode) is annoying and confusing. I strongly suggest following the standard of evaluating with positive numbers the advantage for White and negative for Black always. (What Mephisto did in the eighties is not relevant, and it sux).

I mean, GUIs on PC work like this, no? (SCID, HIARCS, Arena, itd.)


> On 12. jun. 2015, at 12.03, Jürgen Précour <notifications@github.com> wrote:
> 
> There are some issues from ifekali left

- 
Jonas Žnidaršič
jonas@jonas.si

**Comment by djdekker** [comment]
I couldn't agree more with Jonas. Please follow the convention of all chess engines that are in use today (not in the 1980's):
positive evaluation = advantage for White, negative evaluation = advantage for Black.
Otherwise it's getting very confusing!

It would be better still if the score were expressed in pawns instead of centipawns, just like today's desktop computer GUIs do. BTW: my 25 year old Mephisto MMV shows pawn units, not centipawns, as well. ;)

**Comment by LocutusOfPenguin** [comment]
don't misunderstand me wrong!
I dont have any problems if its shown with "white standpoint". But for me - personally -  i dont need that at all. Forgive me, when i develop what i need first (im limited in developing time). esp. since sshivaji wanted to do this #1 himself. => I can spare time this way ;-)

I show what i get from the 2015 chess machine ;-) Just change it...no problem.

**Comment by djdekker** [comment]
No bad feelings :) I really appreciate all the efforts you took to develop Picochess. It's getting better all the time. Thank you very much!

**Comment by puba** [comment]
As usual no big philosophy behind, just an old redneck saying, which means make things as simple as possible, aha and pasulj is a soup, made from beans, very well known and appretiated in ex-Yu countries

**Comment by ifekali** [comment]
Picochess should BEEP on move in all modes (or not as per setting in picochess.ini).


This is especially needed in analysis mode, as one can easily move to fast and the machine WILL HANG soon if this happens (becomes unstable).

Here's a video:

https://youtu.be/vzMDMErBP7A


> On 12. jun. 2015, at 13.55, Jonas Žnidaršič <jonas@jonas.si> wrote:
> 
> As soon it is pulled I will test some

- 
Jonas Žnidaršič
jonas@jonas.si
